[ watchOS ] Crash in PlaybackSessionInterfaceAVKit::PlaybackSessionInterfaceAVKit

jernoble opened the pull request:
#### cecd69451d07f0ab7eb2fbba3661ef19fa085072
<pre>
[ watchOS ] Crash in PlaybackSessionInterfaceAVKit::PlaybackSessionInterfaceAVKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=258750">https://bugs.webkit.org/show_bug.cgi?id=258750</a>
rdar://110387309

Reviewed by Andy Estes and Eric Carlson.

Video playback on watchOS is disabled, but somehow pages are getting far enough along
in video playback to cause a PlaybackSessionInterfaceAVKit to be instantiated. Due to
differences in the implementation of AVKit on watchOS from iOS, our technique of creating
a WebAVPlayerController causes an unimplemented selector exception.

Since we shouldn&apos;t be playing any video on watchOS anyway, just add empty implementations
of VideoFullscreenInterface and PlaybackSessionInterface for use there.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/NullPlaybackSessionInterface.h: Added.
* Source/WebCore/platform/graphics/cocoa/NullVideoFullscreenInterface.h: Added.
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:

Canonical link: <a href="https://commits.webkit.org/265674@main">https://commits.webkit.org/265674@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dad37536b82b7ce9f5d16862d23df7c4d1bf9fcd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11600 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11803 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12165 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13246 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11057 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11621 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14190 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11784 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13928 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11764 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12617 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/9836 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13667 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9909 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10531 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17668 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10982 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13865 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11084 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9134 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10260 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2788 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14538 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10941 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
```diff
@@ -43,7 +43,10 @@
-#if PLATFORM(IOS_FAMILY)
+#if PLATFORM(WATCHOS)
+#include <WebCore/NullVideoFullscreenInterface.h>
+typedef WebCore::NullVideoFullscreenInterface PlatformVideoFullscreenInterface;
```
I know you're just copying the pattern used on other platforms, but would a forward declaration of `NullVideoFullscreenInterface` be sufficient here to define the typedef?

aestes commented on Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
```diff
@@ -38,13 +38,17 @@
 
-#if PLATFORM(IOS_FAMILY)
+#if PLATFORM(WATCHOS)
+typedef WebCore::NullPlaybackSessionInterface PlatformPlaybackSessionInterface;
```
I know you're just copying the pattern used on other platforms, but would a forward declaration of `NullPlaybackSessionInterface` be sufficient here to define the typedef?

eric-carlson reviewed (approved):
Thanks!

jernoble committed:
[ watchOS ] Crash in PlaybackSessionInterfaceAVKit::PlaybackSessionInterfaceAVKit
https://bugs.webkit.org/show_bug.cgi?id=258750
rdar://110387309

Reviewed by Andy Estes and Eric Carlson.

Video playback on watchOS is disabled, but somehow pages are getting far enough along
in video playback to cause a PlaybackSessionInterfaceAVKit to be instantiated. Due to
differences in the implementation of AVKit on watchOS from iOS, our technique of creating
a WebAVPlayerController causes an unimplemented selector exception.

Since we shouldn't be playing any video on watchOS anyway, just add empty implementations
of VideoFullscreenInterface and PlaybackSessionInterface for use there.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/NullPlaybackSessionInterface.h: Added.
* Source/WebCore/platform/graphics/cocoa/NullVideoFullscreenInterface.h: Added.
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:

Canonical link: https://commits.webkit.org/265674@main
